### PR TITLE
fio request encrypt decrypt

### PR DESCRIPTION
**Pull request by elmutt**
encrypt and decrypt fio request content

### CODE REVIEW

**Review by BitHighlander — APPROVED** [review]
:shipit:

File: examples/sandbox/package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "hdwallet-sandbox",
-  "version": "1.3.0",
```
Comment (natven) [inline examples/sandbox/package.json:3]: can this be a minor version bump? breaking changes?

Comment (natven) [reply examples/sandbox/package.json:3]: or patch rather

File: integration/src/fio/fio.ts [hunk 2]
```diff
@@ -1,31 +1,42 @@
 import { bip32ToAddressNList, HDWallet, FioWallet, supportsFio } from "@shapeshiftoss/hdwallet-core";
 
 import { HDWalletInfo } from "@shapeshiftoss/hdwallet-core/src/wallet";
-
 import * as tx01_unsigned from "./tx01.unsigned.json";
 import * as tx02_signed from "./tx02.signed.json";
 
 const MNEMONIC12_NOPIN_NOPASSPHRASE = "alcohol woman abuse must during monitor noble actual mixed trade anger aisle";
+const MNEMONIC12_NOPIN_NOPASSPHRASE2 = "all all all all all all all all all all all all";
```
Comment (natven) [inline integration/src/fio/fio.ts:8]: what you got against alcohol woman abuse?

File: integration/src/fio/fio.ts [hunk 3]
```diff
@@ -1,31 +1,42 @@
...
 
 const TIMEOUT = 60 * 1000;
 
-export function fioTests(get: () => { wallet: HDWallet; info: HDWalletInfo }): void {
+export function fioTests(get: () => { wallet: HDWallet; info: HDWalletInfo; wallet2: HDWallet }): void {
```
Comment (natven) [inline integration/src/fio/fio.ts:12]: confused on wallet2 ❓

Comment (natven) [reply integration/src/fio/fio.ts:12]: ok kept going and answered my question. Simulating a payment request from wallet to another. 👍

Comment (natven) [reply integration/src/fio/fio.ts:12]: maybe call them payerWallet and payeeWallet for clarity?

File: packages/hdwallet-core/src/fio.ts [hunk 4]
```diff
@@ -72,6 +72,14 @@ export interface FioWallet extends FioWalletInfo {
   _supportsFio: boolean;
   fioGetAddress(msg: FioGetAddress): Promise<string>;
   fioSignTx(msg: FioSignTx): Promise<FioSignedTx>;
+  fioDecryptRequestContent(msg: FioEncryptionContent): Promise<string>
+  fioEncryptRequestContent(msg: FioEncryptionContent): Promise<string>
+}
+
+export interface FioEncryptionContent {
```
Comment (natven) [inline packages/hdwallet-core/src/fio.ts:79]: FioRequestContent? totally nitpicking, trying to think what makes it clear what this type is for when it pops up in some engineer's typeahead.

File: packages/hdwallet-native/src/fio.ts [hunk 5]
```diff
@@ -76,5 +81,33 @@ export function MixinNativeFioWallet<TBase extends core.Constructor<NativeHDWall
         signature: res.signatures[0],
       };
     }
-  };
+
+    async fioEncryptRequestContent(msg: core.FioEncryptionContent): Promise<string> {
+      return this.needsMnemonic(!!this.#mnemonic, async () => {
+        const { fioKey: privateKey } = await fio.FIOSDK.createPrivateKeyMnemonic(this.#mnemonic, core.addressNListToBIP32(msg.addressNList))
+        const cipherAlice = fiojs.createSharedCipher({
```
Comment (natven) [inline packages/hdwallet-native/src/fio.ts:88]: is there something significant about Alice?

File: packages/hdwallet-native/src/fio.ts [hunk 6]
```diff
@@ -76,5 +81,33 @@ export function MixinNativeFioWallet<TBase extends core.Constructor<NativeHDWall
...
+          privateKey,
+          publicKey: msg.publicKey,
+          textEncoder,
+          textDecoder
+        })
+        return cipherAlice.encrypt('new_funds_content', msg.content)
```
Comment (natven) [inline packages/hdwallet-native/src/fio.ts:94]: constant for new_funds_content?

File: packages/hdwallet-core/src/fio.ts [hunk 7]
```diff
@@ -72,6 +72,14 @@ export interface FioWallet extends FioWalletInfo {
...
+  addressNList: BIP32Path;
+  content: string;
+  publicKey: string;
```
Comment (natven) [inline packages/hdwallet-core/src/fio.ts:82]: do we need both addressN and pubkey? addressN for payer pubkey for payee?

**Comment by natven** [comment]
Looking good to me. Couple naming things we should address before merge.

File: examples/sandbox/package.json (hunk 1 above)
Comment (elmutt) [reply examples/sandbox/package.json:3]: k

File: packages/hdwallet-native/src/fio.ts (hunk 5 above)
Comment (elmutt) [reply packages/hdwallet-native/src/fio.ts:88]: from example.  will rename

File: packages/hdwallet-native/src/fio.ts (hunk 6 above)
Comment (elmutt) [reply packages/hdwallet-native/src/fio.ts:94]: k

File: lerna.json [hunk 8]
```diff
@@ -1,12 +1,17 @@
 {
   "lerna": "3.22.1",
-  "version": "1.3.0",
+  "version": "1.4.0",
```
Comment (natven) [inline lerna.json:3]: should this be 1.3.1 as well?

File: packages/hdwallet-native/src/fio.ts [hunk 9]
```diff
@@ -2,6 +2,13 @@ import * as core from "@shapeshiftoss/hdwallet-core";
 import * as fio from "fiosdk-offline";
 import fetch, { RequestInfo, RequestInit } from "node-fetch";
 import { NativeHDWalletBase } from "./native";
+import { Fio as fiojs, Ecc as fioecc } from '@fioprotocol/fiojs' // TODO use our forked fioSdk instead of fiojs
```
Comment (natven) [inline packages/hdwallet-native/src/fio.ts:5]: what's this TODO about?

Comment (elmutt) [reply packages/hdwallet-native/src/fio.ts:5]: thing nick talked about in standup

File: lerna.json (hunk 8 above)
Comment (elmutt) [reply lerna.json:3]: sure

**Pull request merged**